include/aws/common/array_list_with_map.h

```diff
@@ -0,0 +1,82 @@
+#ifndef AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+#define AWS_COMMON_ARRAY_LIST_WITH_MAP_H
+
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/common.h>
+#include <aws/common/hash_table.h>
+
+struct aws_array_list_with_map {
+    struct aws_array_list list;
+    struct aws_hash_table map; /* map from the element to the index in the array */
+};
+
+AWS_EXTERN_C_BEGIN
+
+/* TODO: CBMC verification. */
+/**
+ * Initialize the array_list with map structure, which support constant time of insert, remove and get random element
+ * from the data structure.
+ *
+ * The underlying hash map will use hash_fn to compute the hash of each element. equals_fn to compute equality of two
+ * keys.
+ *
+ * @param list_with_map             Pointer of structure to initialize with
+ * @param allocator                 Allocator
+ * @param hash_fn                   Compute the hash of each element
+ * @param equals_fn                 Compute equality of two keys
+ * @param destroy_element_fn        Called when the element is removed
+ * @param item_size                 Size of each element
```

Comment:
There's probably a good reason for this since array_list does this, but why do we make the user give us the size? Can't we just calculate?

Reply:
Good point. Your more familiar with usage so I would defer to you, but I feel like handing the pointer to the user is better since we usually prefer to pass pointers than structs. Again, I would defer to your judgment.